crates/bevy_render/src/render_instances.rs

```diff
@@ -0,0 +1,157 @@
+//! Convenience logic for turning components from the main world into render
+//! instances in the render world.
+//!
+//! This is essentially the same as the `extract_component` module, but
+//! higher-performance because it avoids the ECS overhead.
+
+use std::marker::PhantomData;
+
+use bevy_app::{App, Plugin};
+use bevy_asset::{Asset, AssetId, Handle};
+use bevy_derive::{Deref, DerefMut};
+use bevy_ecs::{
+    component::Component,
+    prelude::Entity,
+    query::{QueryItem, ReadOnlyWorldQuery, WorldQuery},
+    system::{lifetimeless::Read, Query, ResMut, Resource},
+};
+use bevy_utils::EntityHashMap;
+
+use crate::{prelude::ViewVisibility, Extract, ExtractSchedule, RenderApp};
+
+/// Describes how a component gets turned into a render instance for rendering.
+///
+/// Before rendering, a component will be transferred from the main world to the
+/// render world in the [`ExtractSchedule`] step.
+///
+/// This is essentially the same as
+/// [`ExtractComponent`](crate::extract_component::ExtractComponent), but
+/// higher-performance because it avoids the ECS overhead.
+pub trait ExtractToRenderInstance: Component {
```

Comment:
I’m wondering if there is a case for a more flexible version where you give it a query, filter, and instance, and a method as a way of mapping from the query to the instance but it isn’t implemented for a single Component. I’m a bit concerned that doing this per component one wants to extract will be detrimental to performance because it will result in more, independent lookups.